library/alloc/tests/vec.rs

```diff
@@ -1954,3 +1954,25 @@ fn test_vec_swap() {
     assert_eq!(a[0], 42);
     assert_eq!(n, 0);
 }
+
+#[test]
+fn test_extend_from_within_spec() {
+    #[derive(Copy)]
+    struct CopyOnly;
+
+    impl Clone for CopyOnly {
+        fn clone(&self) -> Self {
+            panic!("extend_from_within must use specialization on copy");
+        }
+    }
+
+    vec![CopyOnly, CopyOnly].extend_from_within(..);
+}
+
+#[test]
+fn test_extend_from_within_clone() {
+    let mut v = vec![String::from("sssss"), String::from("12334567890"), String::from("c")];
+    v.extend_from_within(1..);
+
+    assert_eq!(v, ["sssss", "12334567890", "c", "12334567890", "c"]);
```

Comment:
It might be good to add a few other cases, along with some degenerate ones here:

- A complete range
- An out-of-bounds range
- An empty range
- A vec of ZSTs

Reply:
Good call! Actually, the tests allowed me to find unsoundness in the code :sweat_smile: 

I've added tests & fixed unsoundness.